app/ide-desktop/lib/dashboard/src/authentication/src/error.ts

```diff
@@ -0,0 +1,19 @@
+/** @file Contains useful error types common across the module. */
+
+/** An error used to indicate when an unreachable case is hit in a `switch` or `if` statement.
+ *
+ * TypeScript is sometimes unable to determine if we're exhaustively matching in a `switch` or `if`
+ * statement, so we introduce this error in the `default` case (or equivalent) to ensure that we at
+ * least find out at runtime if we've missed a case, or forgotten to update the code when we add a
+ * new case. */
+export class UnreachableCaseError extends Error {
+    constructor(value: never) {
+        /** TypeScript doesn't let us use `never` values in literals, so we have to disable the
+         * rule to allow this. As for why we're using `never` in a literal, it's because we expect
+         * TypeScript to let us know that we're in an unreachable case at type check time (through
+         * the signature of this constructor), but if that fails we still want to know at
+         * runtime. */
+        // eslint-disable-next-line @typescript-eslint/restrict-template-expressions
+        super(`Unreachable case: ${value}`)
```

Comment:
you can do `String(value)` as an alternative way to suppress the error. not sure if it's better though...

Reply:
it *would* reduce the number of `eslint-disables` so it depends whether it's more important for it to be easy to find places where we do something potentially unsafe, or for it to be easy to go through all the eslint-disables instead of them making `eslint-disable`s into a normal occurrence which may make people more willing to `eslint-disable` when they run into a problem (bad!)

... i may be overthinking it though

Reply:
either way the comment seems a little long, i think it can be shorter while still conveying the point - e.g. something like "If we reach this case, `value` is not `never`" (since never is a type containing no values, it cannot exist, but if we reach here something definitely *does* exist)

Reply:
oh whoops instead of `String(value)` you should probably do `JSON.stringify(value)`. it'd error in certain cases (e.g. objects containing references to themselves) but i'd imagine for the most part it'd be better than seeing `"[object Object]"`

Reply:
Good suggestion. Reducing `eslint-disable`s is a good idea, agreed. Will make this change.